### PR TITLE
proxy/iptables: don't call iptables-restore when we don't need to

### PR DESCRIPTION
**Pull request by dcbw**
Diff old and new rules and don't bother calling iptables-restore when nothing has actually changed.

iptables-restore is a very heavy operation and depending on the kernel,
the CPU, and the number of rules to restore, can take a very long time
(~500ms or more).

i7-5600U @ 2.6GHz: 700ms for 1000 services (2+4 cores+threads, kernel 4.6.6)
i7-4790  @ 3.6GHz: 270ms for 1000 services (4+8 cores+threads, kernel 4.6.7)

Other parts of kubernetes (eg kubenet) might be running iptables-restore
too, leading to some pileup as each iptables-restore waits for others to
complete.

Related: https://github.com/kubernetes/kubernetes/issues/26637
Related: https://github.com/kubernetes/kubernetes/issues/33693
Related: https://github.com/kubernetes/kubernetes/pull/35334

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36006)

@thockin @timothysc @danwinship @eparis

### CODE REVIEW

**Comment by dcbw** [comment]
@k8s-bot unit test this https://github.com/kubernetes/kubernetes/pull/36003

**Comment by dcbw** [comment]
Rebase to get revert that fixes https://github.com/kubernetes/kubernetes/issues/35898

File: pkg/util/iptables/iptables.go
Comment (danwinship) [inline pkg/util/iptables/iptables.go]: shouldn't need `n` (likewise below)

File: pkg/util/iptables/iptables.go
Comment (danwinship) [inline pkg/util/iptables/iptables.go]: `return nil, err`? (likewise elsewhere)

File: pkg/proxy/iptables/proxier.go
Comment (danwinship) [inline pkg/proxy/iptables/proxier.go]: possibly in the wrong commit?

File: pkg/proxy/iptables/proxier.go
Comment (danwinship) [inline pkg/proxy/iptables/proxier.go]: There's a difference between oldLines failing to parse and newLines failing to parse: oldLines can fail to parse if the administrator mucks around with iptables behind our back, or if the format of iptables-save changes. But newLines can only fail to parse in the event of a programmer error.

**Comment by dcbw** [comment]
@k8s-bot node e2e test this issue https://github.com/kubernetes/kubernetes/issues/35983

File: pkg/proxy/iptables/proxier.go
Comment (danwinship) [inline pkg/proxy/iptables/proxier.go]: In some failure cases (eg, iptables-save output format changes), every single call to sameIptablesRuleset() will fail... we might want to not log every time?

File: pkg/proxy/iptables/proxier_test.go
Comment (danwinship) [inline pkg/proxy/iptables/proxier_test.go]: not sure what happened there but this isn't gofmt'ed correctly

**Comment by timothysc** [comment]
re-assigning, b/c daniel is out.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Arrays can't be 'const' in Go, something about being able to take the address of an array element which would be illegal for a const.  It also shouldn't really be global since it's used in only one function, so I moved the struct declaration into the function too.

Comment (danwinship) [inline pkg/util/iptables/iptables.go]: But that means it's going to be constructing the array every time you call the function. If you made it global it would only have to do it once.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Fixed, multiple spaces are now handled initially by a helper that splits the line, and empty quoted arguments are passed through as-is (sans quoting of course).

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Fixed.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Yeah.  I just killed them, they were only used in one place in the proxy test code, and only some of them, and I don't want to add more "public" API by expanding them and moving them to iptables.go.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Fixed.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Yeah, fixed.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Fixed (originally copied from somewhere else).

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Fixed for the error cases.

File: pkg/proxy/iptables/proxier.go
Comment (dcbw) [inline pkg/proxy/iptables/proxier.go]: True, fixed.

File: pkg/proxy/iptables/proxier.go
Comment (dcbw) [inline pkg/proxy/iptables/proxier.go]: It'll now only warn if parsing the new rules failed, which should be entirely under our control.

**Comment by dcbw** [comment]
@danwinship PTAL, thanks!

**Comment by dcbw** [comment]
integration test flake is https://github.com/kubernetes/kubernetes/issues/36473

**Comment by dcbw** [comment]
@k8s-bot test this #36473

File: pkg/util/iptables/iptables.go
Comment (danwinship) [inline pkg/util/iptables/iptables.go]: as mentioned above, this has to be re-built every time you call parseRule(), while if it was global it would only have to be built once

Comment (danwinship) [inline pkg/util/iptables/iptables.go]: also maybe sort them into four groups by the optionAttr values, with a comment above each group? (`// arg required, can't negate`, etc)

File: pkg/util/iptables/iptables.go
Comment (danwinship) [inline pkg/util/iptables/iptables.go]: hm, I think there are some more places where you said `%v` when you meant `%q` (across multiple commits)

File: pkg/util/iptables/iptables.go
Comment (danwinship) [inline pkg/util/iptables/iptables.go]: It would simplify the loop if you just did an i++ here like you do for options with arguments. (eg, then both `opt` and `negated` would be local to the loop body)

File: pkg/util/iptables/iptables_test.go
Comment (danwinship) [inline pkg/util/iptables/iptables_test.go]: `err.Error()`. Although none of the test cases sets tc.err anyway

File: pkg/util/iptables/iptables_test.go
Comment (danwinship) [inline pkg/util/iptables/iptables_test.go]: err.Error()

File: pkg/proxy/iptables/proxier.go
Comment (danwinship) [inline pkg/proxy/iptables/proxier.go]: With just a warning, I fear that this is just going to get broken. Someone will add a new rule that the parsing code doesn't know how to parse, no one will notice the warning message, and we'll ship a release where all of this parsing and comparing ends up just being an expensive no-op. OTOH, the code is complicated enough that I'm not comfortable saying we should panic if parsing fails...

Maybe utiliptables.EnsureRule() should return an error if it finds itself about to add a rule that it wouldn't later be able to parse? (Though in that case we'd want to validate module names too, to ensure that there aren't going to be any new "fixups for default module options" needed.)

**Review by danwinship — CHANGES_REQUESTED** [review]

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: I had it that way before, but then we have to also check if i == len(items) a lot.  I'll try it again this way and see.

File: pkg/proxy/iptables/proxier.go
Comment (dcbw) [inline pkg/proxy/iptables/proxier.go]: The proxy testcases should pick up changes to the ruleset though, so it should fail pretty early if syncProxyRules() gets changed.  And it's at least designed to fail safe, as in it shouldn't change existing behavior if things don't work, it'll just always sync and log a message.

Ideally instead of adding textual rules we'd have a more structured iptables module that would use the vishvananda/netlink to read/write stuff and thus be able to more directly compare instead of screen-scraping, but that's a ton of work...

If you really want it I can try the EnsureRule thing and add hostports, userspace proxy, and other testcases too.

Comment (dcbw) [inline pkg/proxy/iptables/proxier.go]: @danwinship the proxy code is the only one so far that uses ParseTableRules(), and while the others could use this too, they don't get called anywhere near as often so they aren't a hotpath like the proxy stuff is.  I'll definitely add comments.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Done.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Done.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Done.

File: pkg/util/iptables/iptables_test.go
Comment (dcbw) [inline pkg/util/iptables/iptables_test.go]: Done.

File: pkg/util/iptables/iptables_test.go
Comment (dcbw) [inline pkg/util/iptables/iptables_test.go]: Done, and added some error testcases.

**Comment by dcbw** [comment]
@danwinship PTAL, thanks

**Comment by dims** [comment]
@thockin @wojtek-t @danwinship @saad-ali : seems a bit big change at the tail end of v1.5. right?

**Comment by saad-ali** [comment]
> @thockin @wojtek-t @danwinship @saad-ali : seems a bit big change at the tail end of v1.5. right?

Agreed, we are well into the code freeze for 1.5. Is this a bug fix? @kubernetes/sig-network for more context. Removing 1.5 milestone until we have verfication.

**Comment by danwinship** [comment]
LGTM

@saad-ali no, this isn't a bugfix, it's a performance improvement

**Comment by jeremyeder** [comment]
@kubernetes/sig-scalability iptables is one of Kube's lowest ceilings in terms of scalability.

**Comment by bprashanth** [comment]
@kubernetes/sig-network is this worth reviewing for 1.5? a number of people have complained about it, so one could view it as a fix to a regression, but the regression was not introduced in 1.4

**Comment by caseydavenport** [comment]
My 2 cents would be to hold off for v1.5 - it stinks that perf isn't great but feels risky to destabilize this late in the cycle.  I do tend to fall on the conservative side for these sorts of things, though.

Is this the sort of thing that could be a candidate for v1.5.1?

**Comment by timothysc** [comment]
I'd consider it important enough.

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: why would this ever happen?

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: This is going to mutate the contents of any string that has more than one consecutive space.  Is that OK?

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: this loop removes the outer quotes but does not unescape embedded quotes or other special characters.

e.g.  

```
# iptables -t nat -D OUTPUT -m comment --comment "foo \"bar\""

# iptables-save | grep foo
-A OUTPUT -m comment --comment "foo \"bar\""
```

you will save `foo \"bar\"`.  I would say either keep the outer quotes or run the whole thing through an unescape

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: Actually, all my other comments be damned, this loop doesn't work, I think:

```
# iptables -t nat -I OUTPUT -m comment --comment "foo \"bar\" baz"

# iptables-save | grep foo
-A OUTPUT -m comment --comment "foo \"bar\" baz"
```

That will defeat your parser in strange ways.

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: yeeuck, this is a race that will hurt to lose...

**Review by thockin — COMMENTED** [review]
I have to run now, will do more later, but I wanted to get this out ASAP.  A couple issues here.

Another issue you have to handle is that iptables will sometimes reorder arguments:

```
# iptables -t nat -I OUTPUT -m tcp -d 1.2.3.4 --dport 9376 -p tcp -j LOG

# sudo iptables-save | grep 1.2.3.4
-A OUTPUT -d 1.2.3.4/32 -p tcp -m tcp --dport 9376 -j LOG
```

Unless you sort the arguments in some canonical order (maybe you do and I missed it while skimming) this is going to be a problem.

Another problem we hit in the past is that different versions of the iptables CLI made different choices wrt this reordering and wrt quoting comments (some older versions did not quote one-word comments.

**Comment by dcbw** [comment]
> Unless you sort the arguments in some canonical order (maybe you do and I missed it while skimming) this is going to be a problem.

@thockin the parsed arguments are stored in a map, so there is no ordering concern.  Ordering also isn't hugely important to iptables itself.  Module names are stored in a list, but that list is sorted before being compared.  So I think this is OK.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: @thockin strings.Split("    ", " ") = [ '', '', '', '' ].  Not particularly useful, but that's a pretty common string split pattern.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: True, though I'd note the proxy code doesn't ever do this.  But I can certainly make that more robust.

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: if you really want to split on whitespace, try `strings.Fields()`, but as described in other commentsI don't think split-on-whitespace is right.

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: this is in `util`.  I expect others to use this, so no time bombs, please :)

I think you simply have to walk through each line, character by character.  When you find a ", you change to "read quoted string" mode until you hit that final unescaped " and change back.  Simple state-machine.

Comment (thockin) [inline pkg/util/iptables/iptables.go]: That "parse a line with quoted strings" function is probably useful on its own.

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: As above I think you simply have to walk through each line, character by character and parse by state machine

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: I mean that this is an arms race with the `iptables` tool - it can add commandline opts arbitrarily, and we need to keep up.  Can we find a better pattern?

Comment (thockin) [inline pkg/util/iptables/iptables.go]: Maybe we can force the caller to pass in a list of flags it uses, so the list becomes site-specific, rather than global? The list below is what kube-proxy needs, but is not generic.

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: Do we actually care about this?  Why not just assume that a `!` always applies to the next arg.  If if doesn't work as an actual iptables command, who cares.  That simplifies this and more importantly the table of commands can become a map of `flagName : hasArg`

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: You'd have to handle -I, -X, etc to be complete.  but...

Why bother to semantically parse?  Why not just parse it into a set of flags (and flag-arg pairs) and totally ignore any meaning?  Is that really bringing enough value to justify?

File: pkg/util/iptables/iptables.go
Comment (thockin) [inline pkg/util/iptables/iptables.go]: I don't understand these?  It does raise another issue - there are many things that the iptables command allows that get mutated on readback.  --mask becomes --xmask, and the hex value of the mask gets canonicalized.  If we can't handle that all, I am reluctant to embed this logic in a generic location - people will assume it works.

kube-proxy pre-canonicalizes mostly for historical reasons, but it would be really easy to slip up on that (or to have drift across tool versions).  Since this is distant from the use-site, it will be very hard to find.

**Review by thockin — CHANGES_REQUESTED** [review]
This is the first commit only.

File: pkg/proxy/iptables/proxier.go
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: This is not right.  At least on my machines, it is actually best-fit:

```
-A FOO -j MARK --set-xmark 0x1/0x1
-A FOO -j MARK --set-xmark 0x10/0x10
-A FOO -j MARK --set-xmark 0x100/0x100
-A FOO -j MARK --set-xmark 0x1000/0x1000
-A FOO -j MARK --set-xmark 0x10000/0x10000
-A FOO -j MARK --set-xmark 0x100000/0x100000
-A FOO -j MARK --set-xmark 0x1000000/0x1000000
-A FOO -j MARK --set-xmark 0x10000000/0x10000000
```

File: pkg/proxy/iptables/proxier.go
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: Can we peel off little changes like this as separate PRs or commits?

File: pkg/proxy/iptables/proxier.go
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: We should be EOL'ing this anyway, since kubelet now installs the mark chains..

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: Note that iptables-save only ever outputs "-A" lines...  but we need to parse at least the top-level commands that specify which chain this line is for, since we need to be able to ignore chains that aren't owned by kube.  I'll remove the specific parsing of the others though.

File: pkg/util/iptables/iptables.go
Comment (dcbw) [inline pkg/util/iptables/iptables.go]: @thockin would you prefer we pre-canonicalize everywhere?  Easy enough to do, but kinda fragile.  At least the failure case is to safe; always rewrite which is what we do now.

An alternative would be to stop constructing pure iptables-restore lines, but add actual commands to pkg/util/iptables to start a new rule and add options.  Then we can validate them and ensure they will pass our match filters. Sound OK?

Comment (thockin) [inline pkg/util/iptables/iptables.go]: We can't pre-canonicalize entirely.  kube-proxy took pain to ensure it feeds args in the same order and format that relatively modern iptables reads it back, but as you saw in that other PR, older iptables is different, and I'll bet 100 bucks it changes again at some point in the future (or we get it wrong on our input).  My point is mostly - if we're going to do this, we should decide - are we generic & dumb (with opaque flags and args), or generic & smart (implementing a parser for all known iptables flags), or specialized and smart (implement a parser for all flags we know we need)?  We shouldn't be halfway.

I'm not sure I understand your alternative?

**Comment by thockin** [comment]
What conditions would cause us to have a significant number of no-op writes?

**Comment by timothysc** [comment]
> What conditions would cause us to have a significant number of no-op writes? 

Multiple HA'd components using endpoint annotations.

**Comment by dcbw** [comment]
> This is not right. At least on my machines, it is actually best-fit:

@thockin %0yX is supposed to zero-pad the front out to the specified number of characters.  So I don't see how "%08X" wouldn't result in 0x00000001...  Best fit is entirely contrary to the definition of "%0".  In any case with Go 1.7 on my machine, I do get zero padding.  You're saying you don't?

**Comment by dcbw** [comment]
@thockin I think I've addressed all your comments except about the default module option fixups.  I'll keep thinking about that.

Turns out we don't have to hardcode options at all, since AFAIK we can just use - as a marker for option and then take whatever's next as the argument.  I've also redone the line parsing as you suggest, and removed the special handling of Comment and Jump arguments, folding them into the options dict.

**Comment by dcbw** [comment]
@k8s-bot gci gce e2e test this issue https://github.com/kubernetes/kubernetes/issues/37361

**Comment by jeremyeder** [comment]
To give you a sense of what we're seeing wrt iptables save/restore churn, here is a graph showing CPU utilization across a variety of system roles in a cluster.  Avg in the 2+ core range for iptables rule validation.

This cluster has 1000 nodes and 14,000 pods.  At scales far lower than this, you can find iptables consuming a core or so on every node in the cluster at any given time.  For RH'rs reading, if you'd like access please ping me internally.

iptables

**Comment by warden** [comment]
@jeremyeder out of curiosity, how many services do you have in the cluster?

**Comment by jeremyeder** [comment]
Approximately 600 services.

**Comment by widgetpl** [comment]
@jeremyeder do you have any network issues like packet drops etc. couse we have faced such issues when we hit to around 500 services.

**Comment by jeremyeder** [comment]
@widgetpl Just confirmed, no drops or errors on any of the interfaces.

**Comment by dcbw** [comment]
Going to close this one too, as the core issue may have already been addressed by other changes like https://github.com/kubernetes/kubernetes/pull/41022 and https://github.com/kubernetes/kubernetes/pull/38996

**Pull request closed without merging**
